### PR TITLE
MONGOID-5608 allow using `#exists?` with args on relations

### PR DESCRIPTION
**Pull request by jamis**
Adds extended support for `#exists?` on `embeds_many` and `has_many` associations, so the API is the same as is present elsewhere in Mongoid, where the `#exists?` accepts conditions that can be used to narrow the scope of the query.

### CODE REVIEW

File: lib/mongoid/association/referenced/has_many/proxy.rb [hunk 1]
```diff
@@ -217,9 +217,23 @@ def each(&block)
...
           # @return [ true | false ] True is persisted documents exist, false if not.
-          def exists?
-            criteria.exists?
+          def exists?(id_or_conditions = :none)
+            case id_or_conditions
+            when :none then criteria.exists?
+            when nil, false then false
+            when Hash then criteria.where(id_or_conditions).exists?
+            else criteria.where(_id: id_or_conditions).exists?
+            end
```
Comment (johnnyshields) [inline lib/mongoid/association/referenced/has_many/proxy.rb:236]: why can't this delegate to criteria#exists?

Comment (jamis) [reply lib/mongoid/association/referenced/has_many/proxy.rb:236]: Good catch! It certainly could. I'll fix.

File: lib/mongoid/association/referenced/has_many/proxy.rb [hunk 2]
```diff
@@ -217,9 +217,18 @@ def each(&block)
...
+          #   conditions, and returns true if any match. Any other argument is
+          #   interpreted as an id, and queries for the existence of documents
+          #   in the association with a matching _id.
+          #
           # @return [ true | false ] True is persisted documents exist, false if not.
-          def exists?
-            criteria.exists?
+          def exists?(id_or_conditions = :none)
+            criteria.exists?(id_or_conditions)
           end
```
Comment (johnnyshields) [inline lib/mongoid/association/referenced/has_many/proxy.rb:232]: Why not:

```
delegate :exists?, to: :criteria
```

File: lib/mongoid/association/referenced/has_many/proxy.rb [hunk 3]
```diff
@@ -217,9 +217,18 @@ def each(&block)
...
+          #   will always return false. When a Hash is given, this queries
+          #   the documents in the association for those that match the given
+          #   conditions, and returns true if any match. Any other argument is
+          #   interpreted as an id, and queries for the existence of documents
+          #   in the association with a matching _id.
+          #
           # @return [ true | false ] True is persisted documents exist, false if not.
-          def exists?
-            criteria.exists?
+          def exists?(id_or_conditions = :none)
```
Comment (johnnyshields) [inline lib/mongoid/association/referenced/has_many/proxy.rb:230]: Why not:

```
delegate :exists?, to: :criteria
```

Comment (jamis) [reply lib/mongoid/association/referenced/has_many/proxy.rb:230]: That would work, too. In general I'm not a fan of that style. It tends to stymie investigation into execution flow, at least for me. I'll think about it, though.

Comment (johnnyshields) [reply lib/mongoid/association/referenced/has_many/proxy.rb:230]: It is recommended by Rubocop: https://docs.rubocop.org/rubocop-rails/cops_rails.html#railsdelegate

**Review by comandeo — APPROVED** [review]
LGTM!
We have now tests for `:none` and `Hash`. Should we also add tests for `nil` and `false` ?

**Review by comandeo — APPROVED** [review]
